src/dist/replication/lib/mutation_log.cpp

```diff
@@ -852,67 +851,111 @@ std::pair<log_file_ptr, int64_t> mutation_log::mark_new_offset(size_t size,
     return std::make_pair(_current_log_file, write_start_offset);
 }
 
-/*static*/ error_code mutation_log::replay(log_file_ptr log,
-                                           replay_callback callback,
-                                           /*out*/ int64_t &end_offset)
+namespace internal {
+
+inline static error_s read_log_block(log_file_ptr &log,
+                                     int64_t &end_offset,
+                                     std::unique_ptr<binary_reader> &reader,
+                                     blob &bb)
 {
-    end_offset = log->start_offset();
-    ddebug("start to replay mutation log %s, offset = [%" PRId64 ", %" PRId64 "), size = %" PRId64,
-           log->path().c_str(),
-           log->start_offset(),
-           log->end_offset(),
-           log->end_offset() - log->start_offset());
+    FAIL_POINT_INJECT_F("mutation_log_read_log_block", [](string_view) -> error_s {
+        return error_s::make(ERR_INCOMPLETE_DATA, "mutation_log_read_log_block");
+    });
 
-    ::dsn::blob bb;
-    log->reset_stream();
     error_code err = log->read_next_log_block(bb);
     if (err != ERR_OK) {
-        return err;
+        return error_s::make(err, "failed to read log block");
     }
-
-    std::shared_ptr<binary_reader> reader(new binary_reader(std::move(bb)));
+    reader = dsn::make_unique<binary_reader>(bb);
     end_offset += sizeof(log_block_header);
 
-    // read file header
-    end_offset += log->read_file_header(*reader);
-    if (!log->is_right_header()) {
-        return ERR_INVALID_DATA;
+    return error_s::ok();
+}
+
+} // namespace internal
+
+/*static*/ error_s mutation_log::replay_block(log_file_ptr &log,
+                                              replay_callback &callback,
+                                              size_t start_offset,
+                                              int64_t &end_offset)
+{
+    FAIL_POINT_INJECT_F("mutation_log_replay_block", [](string_view) -> error_s {
+        return error_s::make(ERR_INCOMPLETE_DATA, "mutation_log_replay_block");
+    });
+
+    blob bb;
+    std::unique_ptr<binary_reader> reader;
+
+    // read from start
+    int64_t global_start_offset = start_offset + log->start_offset();
+    if (global_start_offset != end_offset) {
+        end_offset = global_start_offset;
     }
+    log->reset_stream(start_offset);
```

Comment:
1. global_start_offset变量名，感觉有些奇怪，block_start_offset是不是更好？就是表示这次block的start_offset
2. 参照我下面的建议，在调用replay_block前将end_offset设置为0, 在调用中将它初始化，然后在read的时候再增加，这样就不用做global_start_offset和end_offset的比较，因为我们并不在意传入的end_offset的值
3. 为什么每次调用的时候要reset_stream呢？stream的read_next已经维护了offset，不调用reset它就是下次读取的位置

Reply:
3. 这是好问题，在这里 reset_stream 是一个很重要的操作，因为热备份读日志的时候经常会出现读到文件末尾，发现文件还没写完导致读失败，或者这一小段读文件出错。这时候你不能从文件头开始读，这样很不好，你要重新读刚才读失败的这一小段。这个事情在实验之后效果很好。

所以在失败的时候我们会从上一次读的地方接着往下读，也就是 start_offset，这个是用户维护的，而 stream 自己维护的 offset 可能不准。通常来讲，如果一切都正常的话，stream 自己维护的 offset 就是 start_offset。

Reply:
建议把这个3的回答加到上面的描述或者代码中，不然这个conversation resolve了就不容易看到了

Reply:
>  However, for duplication, the replay of mutation_log becomes a block by block operation. Because it will impose much pressure on this replica to read the entire file once.

Reply:
这句话就是之前有歧义的标题额，replay log一直都是by block的，并不是为了热备功能修改的，我希望你补充的是reset_stream这个函数，你为什么要添加offset作为参数，就像你上面给我的解释一样